### DISCUSSION TITLE
WebRTC delay building up with time

### DISCUSSION THREAD

**Discussion by JankiMandviya** [aiortc/aiortc#1238] [Q&A]
I am using aiortc library to stream output of my deep learning algorithm to web browser.
I have been using following example:
https://github.com/aiortc/aiortc/tree/main/examples/webcam
When i am streaming the webcam output from the source /dev/video0 streaming performance is really good.
But to apply opencv operations and my object detection algorithm i am using VideoStreamTrack class, it is resulting into very high latency in streaming and delay builds up with time. It starts from some where between 4-5 seconds and ends up being upto 30 seconds. 
How can i solve this issue.
Note: object detection algorithm takes only 30-40ms on my system.

**Comment by dodobyte** [comment]
you can't block VideoStreamTrack, run opencv ops from a separate thread and use a queue to send images, consume from track if they're ready.

**Comment by JankiMandviya** [comment]
> you can't block VideoStreamTrack, run opencv ops from a separate thread and use a queue to send images, consume from track if they're ready.

you mean. My main code will start the server as shown in the examples provided by aiortc. Then i should make a thread to capture and process the frames and put those frames in queue, and then i will fetch output images in VideoStreamTrack class? 
Am i correct?

**Comment by dodobyte** [comment]
yes, I'd recommend to use deque and in case you're producing frames slower than desired fps, don't popleft the last frame from the video track, get it with queue[0].

**Comment by JankiMandviya** [comment]
okay thank you.

**Comment by JankiMandviya** [comment]
```
import argparse
import asyncio
import json
import logging
import os
import platform
import ssl
import cv2
import numpy as np
from collections import deque
import threading
from aiohttp import web
from aiortc import RTCPeerConnection, RTCSessionDescription , VideoStreamTrack
from aiortc.contrib.media import MediaPlayer, MediaRelay
from aiortc.rtcrtpsender import RTCRtpSender
import av

ROOT = os.path.dirname(__file__)


relay = None
webcam = None
queue = deque(maxlen=60)

def cap_and_process():
    cap = cv2.VideoCapture('http://<IP_add_of_IPWebcam>/cap')
    cap.open('http://<IP_add_of_IPWebcam>/video')
    
    # Read frame from OpenCV
    while True:
        ret, frame = cap.read()
        if not ret:
            return None
        frame_rgb = cv2.cvtColor(frame,cv2.COLOR_BGR2RGB)
        if not queue.full():
            print("filling the queue")
            queue.append(frame_rgb)

class OpenCVCapture(VideoStreamTrack):
    def __init__(self,frame_deque):
        super().__init__()  # VideoStreamTrack is a subclass of MediaStreamTrack
        self.frame_deque = frame_deque

    async def recv(self):
        pts, time_base = await self.next_timestamp()
        
        if self.frame_deque:
            frame = self.frame_deque.popleft()
        
        video_frame = av.VideoFrame.from_ndarray(frame, format="rgb24")
        video_frame.pts = pts
        video_frame.time_base = time_base

        return video_frame
    

def create_local_tracks(play_from, decode):
    global relay, webcam

    if play_from:
        player = MediaPlayer(play_from, decode=decode)
        return player.audio, player.video
    else:
        options = {"framerate": "30", "video_size": "1280x720"}
        if relay is None:
            webcam = OpenCVCapture(frame_deque = queue)  # 0 is the default webcam
            relay = MediaRelay()
            print(dir(webcam))
        return None, relay.subscribe(webcam)


def force_codec(pc, sender, forced_codec):
    kind = forced_codec.split("/")[0]
    codecs = RTCRtpSender.getCapabilities(kind).codecs
    transceiver = next(t for t in pc.getTransceivers() if t.sender == sender)
    transceiver.setCodecPreferences(
        [codec for codec in codecs if codec.mimeType == forced_codec]
    )


async def index(request):
    content = open(os.path.join(ROOT, "index.html"), "r").read()
    return web.Response(content_type="text/html", text=content)


async def javascript(request):
    content = open(os.path.join(ROOT, "client.js"), "r").read()
    return web.Response(content_type="application/javascript", text=content)


async def offer(request):
    params = await request.json()
    offer = RTCSessionDescription(sdp=params["sdp"], type=params["type"])

    pc = RTCPeerConnection()
    pcs.add(pc)

    @pc.on("connectionstatechange")
    async def on_connectionstatechange():
        print("Connection state is %s" % pc.connectionState)
        if pc.connectionState == "failed":
            await pc.close()
            pcs.discard(pc)

    # open media source
    audio, video = create_local_tracks(
        args.play_from, decode=not args.play_without_decoding
    )

    if audio:
        audio_sender = pc.addTrack(audio)
        if args.audio_codec:
            force_codec(pc, audio_sender, args.audio_codec)
        elif args.play_without_decoding:
            raise Exception("You must specify the audio codec using --audio-codec")

    if video:
        video_sender = pc.addTrack(video)
        if args.video_codec:
            force_codec(pc, video_sender, args.video_codec)
        elif args.play_without_decoding:
            raise Exception("You must specify the video codec using --video-codec")

    await pc.setRemoteDescription(offer)

    answer = await pc.createAnswer()
    await pc.setLocalDescription(answer)

    return web.Response(
        content_type="application/json",
        text=json.dumps(
            {"sdp": pc.localDescription.sdp, "type": pc.localDescription.type}
        ),
    )


pcs = set()


async def on_shutdown(app):
    # close peer connections
    coros = [pc.close() for pc in pcs]
    await asyncio.gather(*coros)
    pcs.clear()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="WebRTC webcam demo")
    parser.add_argument("--cert-file", help="SSL certificate file (for HTTPS)")
    parser.add_argument("--key-file", help="SSL key file (for HTTPS)")
    parser.add_argument("--play-from", help="Read the media from a file and sent it.")
    parser.add_argument(
        "--play-without-decoding",
        help=(
            "Read the media without decoding it (experimental). "
            "For now it only works with an MPEGTS container with only H.264 video."
        ),
        action="store_true",
    )
    parser.add_argument(
        "--host", default="0.0.0.0", help="Host for HTTP server (default: 0.0.0.0)"
    )
    parser.add_argument(
        "--port", type=int, default=8080, help="Port for HTTP server (default: 8080)"
    )
    parser.add_argument("--verbose", "-v", action="count")
    parser.add_argument(
        "--audio-codec", help="Force a specific audio codec (e.g. audio/opus)"
    )
    parser.add_argument(
        "--video-codec", help="Force a specific video codec (e.g. video/H264)"
    )

    args = parser.parse_args()

    if args.verbose:
        logging.basicConfig(level=logging.DEBUG)
    else:
        logging.basicConfig(level=logging.INFO)

    if args.cert_file:
        ssl_context = ssl.SSLContext()
        ssl_context.load_cert_chain(args.cert_file, args.key_file)
    else:
        ssl_context = None
    p_thread = threading.Thread(target = cap_and_process,daemon=True)
    p_thread.start()

    app = web.Application()
    app.on_shutdown.append(on_shutdown)
    app.router.add_get("/", index)
    app.router.add_get("/client.js", javascript)
    app.router.add_post("/offer", offer)
    web.run_app(app, host=args.host, port=args.port, ssl_context=ssl_context)
```

This is my code after making changes according to your suggestion. 
But still delay increases with time. 
Note : here i have not applied the object detection algorithm. I am trying simple processing to check the delay of base code.

**Comment by EricVoll** [comment]
What happened to me once in a similar case was that the queue was too long and built up a buffer because the track was slower at consuming images than the camera-thread at placing the images in the queue. 
The solution was to drop some frames by effectively setting the queue size to 1. At that point you don't need a queue anymore though.

**Comment by JankiMandviya** [comment]
> What happened to me once in a similar case was that the queue was too long and built up a buffer because the track was slower at consuming images than the camera-thread at placing the images in the queue. 
> The solution was to drop some frames by effectively setting the queue size to 1. At that point you don't need a queue anymore though.

I tried a frame-dropping approach as well, but even taking only one frame per second (the camera has a frame rate of 30 fps) and dropping the rest, I can not beat the slower tracking rate. I can not afford to drop many frames because it will make video streaming seem glitchy and lag.

**Comment by EricVoll** [comment]
Huh interesting. For us we're running a WebRTC Service with aiortc for two cameras at 15fps via the internet without any issues. We're throttling because of our cameras, and not because of aiortc.

1 fps should _easily_ be possible.

**Comment by JankiMandviya** [comment]
> Huh interesting. For us we're running a WebRTC Service with aiortc for two cameras at 15fps via the internet without any issues. We're throttling because of our cameras, and not because of aiortc.
> 
> 1 fps should _easily_ be possible.

Are you doing any OpenCV operations on camera feed before streaming it to WebRTC?

**Comment by EricVoll** [comment]
Only rendering some text into the camera image. But I'm very sure that in your case the problem isn't opencv but something with the webrtc implementation. You can try to just create a black image via numpy and stream that via webrtc. If it's still so slow, then the problem is webrtc, not open cv.

**Comment by JankiMandviya** [comment]
> Only rendering some text into the camera image. But I'm very sure that in your case the problem isn't opencv but something with the webrtc implementation. You can try to just create a black image via numpy and stream that via webrtc. If it's still so slow, then the problem is webrtc, not open cv.

Okay, I will try that. Thank you so much!

**Comment by shenwpo** [comment]
You can try replacing `frame = self.frame_deque.popleft()`
with `frame = await asyncio.to_thread(self.frame_deque.popleft)`
